src/python/gudhi/hera/wasserstein.cc

```diff
@@ -8,29 +8,126 @@
  *      - YYYY/MM Author: Description of the modification
  */
 
-#include <wasserstein.h> // Hera
-
 #include <pybind11_diagram_utils.h>
 
-double wasserstein_distance(
+#ifdef _MSC_VER
+// https://github.com/grey-narn/hera/issues/3
+// ssize_t is a non-standard type (well, posix)
+using py::ssize_t;
+#endif
+
+#include <hera/wasserstein.h>
+#include <gudhi/Debug_utils.h>
+
+// Unlike bottleneck, for wasserstein, we need to add the index ourselves (if we want the matching)
+static auto make_hera_point(double x, double y, py::ssize_t i) { return hera::DiagramPoint<double>(x, y, i); };
+
+py::object wasserstein_distance(
     Dgm d1, Dgm d2,
     double wasserstein_power, double internal_p,
-    double delta)
+    double delta, bool return_matching)
 {
   // I *think* the call to request() has to be before releasing the GIL.
-  auto diag1 = numpy_to_range_of_pairs(d1);
-  auto diag2 = numpy_to_range_of_pairs(d2);
-
-  py::gil_scoped_release release;
-
-  hera::AuctionParams<double> params;
-  params.wasserstein_power = wasserstein_power;
-  // hera encodes infinity as -1...
-  if(std::isinf(internal_p)) internal_p = hera::get_infinity<double>();
-  params.internal_p = internal_p;
-  params.delta = delta;
-  // The extra parameters are purposely not exposed for now.
-  return hera::wasserstein_dist(diag1, diag2, params);
+  auto diag1 = numpy_to_range_of_pairs(d1, make_hera_point);
+  auto diag2 = numpy_to_range_of_pairs(d2, make_hera_point);
+  int n1 = boost::size(diag1);
+  int n2 = boost::size(diag2);
+  hera::AuctionResult<double> res;
+  double dist;
+
+  { // No Python allowed in this section
+    py::gil_scoped_release release;
+
+    hera::AuctionParams<double> params;
+    params.wasserstein_power = wasserstein_power;
+    // hera encodes infinity as -1...
+    if(std::isinf(internal_p)) internal_p = hera::get_infinity<double>();
+    params.internal_p = internal_p;
+    params.delta = delta;
+    if(return_matching) {
+      params.return_matching = true;
+      params.match_inf_points = true;
+    }
+    // The extra parameters are purposely not exposed for now.
+    res = hera::wasserstein_cost_detailed(diag1, diag2, params);
+    dist = std::pow(res.cost, 1./params.wasserstein_power);
+  }
+
+  if(!return_matching)
+    return py::cast(dist);
+
+  if(dist == std::numeric_limits<double>::infinity())
+    return py::make_tuple(dist, py::none());
+
+  // bug in Hera, matching_a_to_b_ is empty if one diagram is empty or both diagrams contain the same points
```

Comment:
Could you please open an [issue on hera](https://github.com/anigmetov/hera/issues) to track this ?

Reply:
There is already a TODO in Hera's code:
https://github.com/anigmetov/hera/blob/8bfdd4bd32f005c18b5c75c502b987de552d6e48/include/hera/wasserstein.h#L286